### DISCUSSION TITLE
How to use transform for Model.create and update?

### DISCUSSION THREAD

**Discussion by phstc** [sensedeep/dynamodb-onetable#332] [General]
I was trying to implement a `transform` when creating and updating.

```typescript
  const transform = (
    model: AnyModel,
    op: string,
    name: string,
    value: any,
    properties: OneProperties,
  ): any => {
    return ...; // <====== what should I return here?
  };

  await Test.update(
    { name: 'testxd', id: '8', random: String(Math.random()) },
    { transform },
  );
```

But that's failing with `(node:54912) UnhandledPromiseRejectionWarning: OneTableError: Cannot find existing item to update...` 

1. What's the expected return value for `transform`? The TypeScript interface is defined with `any`
2. Is `transform` going to be deprecated? `WARNING: params.transform functions are DEPRECATED and will be removed soon.`
3. I was planning to use `transform` for adding additional attributes to unique records `_unique`. So I could also retrieve additional data querying when `_unique` records. [Currently, it saves only `pk` and `sk`](https://github.com/sensedeep/dynamodb-onetable/blob/23395dbb17b66662098f5a40b2838093d9d8c17d/src/Model.js#L610-L613). I thought of a transform for additional columns. Is it a feasible idea?

**Comment by mobsense** [comment]
Your error may be unrelated to the transform. Use params {log: true} to see the actual request.

You should return any value of the correct type for that attribute from a transform.  Any is right, because the type can be multiple types depending on the attribute being transformed.

The deprecation warning is incorrect -- I've removed it.

Regarding adding additional fields in a transform ... the params transform is meant for simply transforming values, not for adding new fields. The table transform works on the entire record, so you may be able to use that (I've not tested that use case).  

Typically we recommend having attributes defined in the schema -- why do you want/need to add them to the unique record?

**Comment by phstc** [comment]
Thank you!

> why do you want/need to add them to the unique record?

It is a way to avoid a GSI for specific use cases. For example, if in addition to the unique record pk & sk I can add, let's say `accountId`, I could query for the unique pk & sk and retrieve the `accountId`. 

Currently, to accomplish that, I need to have the unique record for the unique attribute, and I need to set this unique attribute as a GSI in the original record so that I can query based on it.

**Comment by mobsense** [comment]
Why don't you just retrieve the full item?  It would contain accountId?

**Comment by phstc** [comment]
> Why don't you just retrieve the full item? It would contain accountId?

@mobsense because I would need to add a GSI for querying by `name`, for example, if I have a schema like this:

```typescript
const Schema = {
  format: 'onetable:1.1.0',
  version: '0.0.1',
  indexes: {
    primary: { hash: 'pk', sort: 'sk' },
  },
  models: {
    Test: {
      pk: { type: String, value: '${id}' },
      sk: { type: String, value: 'test' },
      id: { type: String, required: true },
      name: { type: String, unique: true },
    },
  } as const,
  params: {
    isoDates: true,
    timestamps: true,
  },
};
```

I can't query records by `name` (which is `unique: true` and has a row for it). Since we are inserting a row for the `name` uniqueness, I could just query for that instead, for example `Test.get({ pk: "_unique#Test#name"})` and I have use cases that I don't need the full item, if I could save one or two additional attributes for the unique records, I would be fine querying for the that alone.

Without that, I need to add a GSI:

```typescript
const Schema = {
  format: 'onetable:1.1.0',
  version: '0.0.1',
  indexes: {
    primary: { hash: 'pk', sort: 'sk' },
    gsi: { hash: 'gsipk', sort: 'gsisk },
  },
  models: {
    Test: {
      pk: { type: String, value: '${id}' },
      sk: { type: String, value: 'test' },
      gsipk: { type: String, value: '${name}' },
      gsisk: { type: String, value: 'test' },
      id: { type: String, required: true },
      name: { type: String, unique: true },
    },
  } as const,
  params: {
    isoDates: true,
    timestamps: true,
  },
};
```

Then `Test.get({ name: "name"}, { index: 'gsi' })`.

Does that make sense?
